File: tests/test-qwen-integration.py
#!/usr/bin/env python3
"""
Test script for Qwen3-Coder 480B integration with MaiFarm
Tests the LLM proxy, API integration, and multi-agent orchestration
"""

import os
import sys
import json
import time
import subprocess
import requests
import asyncio
from pathlib import Path
from typing import Dict, Any, List

# Color codes for output
GREEN = '\033[92m'
YELLOW = '\033[93m'
RED = '\033[91m'
RESET = '\033[0m'
BOLD = '\033[1m'

def print_status(message: str, status: str = "info"):
    """Print colored status messages"""
    if status == "success":
        print(f"{GREEN}✓{RESET} {message}")
    elif status == "warning":
        print(f"{YELLOW}⚠{RESET} {message}")
    elif status == "error":
        print(f"{RED}✗{RESET} {message}")
    else:
        print(f"  {message}")

def check_environment():
    """Check environment configuration for Qwen"""
    print(f"\n{BOLD}=== Environment Configuration ==={RESET}")
    
    required_vars = {
        "QWEN_API_KEY": "DashScope API key for Qwen3-Coder",
        "QWEN_API_ENDPOINT": "DashScope API endpoint",
        "QWEN_MODEL": "Qwen model name (e.g., qwen-coder-480b)",
        "LLM_PROXY_URL": "LLM proxy server URL",
        "USE_LLM_PROXY": "Whether to use LLM proxy"
    }
    
    all_configured = True
    for var, description in required_vars.items():
        value = os.getenv(var, "")
        if value:
            # Mask API key for security
            if "KEY" in var:
                display_value = value[:10] + "..." if len(value) > 10 else "***"
            else:
                display_value = value
            print_status(f"{var}: {display_value}", "success")
        else:
            print_status(f"{var}: Not configured ({description})", "warning")
            if var in ["QWEN_API_KEY"]:
                all_configured = False
    
    return all_configured

def test_llm_proxy():
    """Test LLM proxy server"""
    print(f"\n{BOLD}=== Testing LLM Proxy ==={RESET}")
    
    proxy_url = os.getenv("LLM_PROXY_URL", "http://localhost:8001")
    
    # Check if proxy is running
    try:
        response = requests.get(f"{proxy_url}/providers", timeout=5)
        if response.ok:
            providers = response.json().get("providers", [])
            print_status(f"LLM Proxy is running at {proxy_url}", "success")
            print_status(f"Available providers: {', '.join(providers)}", "info")
            
            if "qwen" in providers:
                print_status("Qwen provider is available", "success")
                return True
            else:
                print_status("Qwen provider not available in proxy", "warning")
                return False
        else:
            print_status(f"LLM Proxy returned error: {response.status_code}", "error")
            return False
    except requests.exceptions.RequestException as e:
        print_status(f"LLM Proxy not accessible: {e}", "error")
        print_status("Starting LLM proxy server...", "info")
        
        # Try to start the proxy
        try:
            subprocess.Popen(
                ["python", "llm_proxy.py", "--proxy"],
                stdout=subprocess.DEVNULL,
                stderr=subprocess.DEVNULL
            )
            time.sleep(3)  # Wait for proxy to start
            
            # Try again
            response = requests.get(f"{proxy_url}/providers", timeout=5)
            if response.ok:
                print_status("LLM Proxy started successfully", "success")
                return True
        except Exception as start_error:
            print_status(f"Failed to start LLM proxy: {start_error}", "error")
            return False
    
    return False

def test_qwen_api():
    """Test direct Qwen API via proxy"""
    print(f"\n{BOLD}=== Testing Qwen3-Coder API ==={RESET}")
    
    proxy_url = os.getenv("LLM_PROXY_URL", "http://localhost:8001")
    
    test_messages = [
        {"role": "user", "content": "Write a simple Python function to calculate factorial"}
    ]
    
    request_data = {
        "model": "qwen/qwen-coder-480b",
        "messages": test_messages,
        "max_tokens": 200,
        "temperature": 0.7
    }
    
    try:
        print_status("Sending test request to Qwen3-Coder...", "info")
        response = requests.post(
            f"{proxy_url}/chat/completions",
            json=request_data,
            timeout=30
        )
        
        if response.ok:
            result = response.json()
            if "choices" in result and len(result["choices"]) > 0:
                content = result["choices"][0]["message"]["content"]
                print_status("Qwen3-Coder responded successfully", "success")
                print_status(f"Response preview: {content[:100]}...", "info")
                
                # Check if response contains code
                if "def" in content or "function" in content:
                    print_status("Response contains code as expected", "success")
                return True
            else:
                print_status("Unexpected response format", "error")
                print(json.dumps(result, indent=2))
                return False
        else:
            print_status(f"API request failed: {response.status_code}", "error")
            print_status(f"Error: {response.text}", "error")
            return False
            
    except requests.exceptions.RequestException as e:
        print_status(f"API request failed: {e}", "error")
        return False

def test_ollama_local():
    """Test local Ollama Qwen models"""
    print(f"\n{BOLD}=== Testing Local Ollama Qwen Models ==={RESET}")
    
    # Check if Ollama is installed
    ollama_check = subprocess.run(
        "which ollama",
        shell=True,
        capture_output=True,
        text=True
    )
    
    if ollama_check.returncode != 0:
        print_status("Ollama not installed", "warning")
        print_status("Install from: https://ollama.ai", "info")
        return False
    
    print_status("Ollama is installed", "success")
    
    # Check for Qwen models
    list_result = subprocess.run(
        "ollama list | grep -i qwen",
        shell=True,
        capture_output=True,
        text=True
    )
    
    if list_result.returncode == 0 and list_result.stdout:
        models = list_result.stdout.strip().split('\n')
        print_status(f"Found {len(models)} Qwen model(s):", "success")
        for model in models:
            print_status(f"  - {model.split()[0]}", "info")
        return True
    else:
        print_status("No Qwen models found locally", "warning")
        print_status("Pull a model with: ollama pull qwen2.5-coder:7b", "info")
        return False

def test_multi_agent():
    """Test multi-agent orchestration with Qwen"""
    print(f"\n{BOLD}=== Testing Multi-Agent Orchestration ==={RESET}")
    
    # Check if orchestrator.py exists
    if not Path("orchestrator.py").exists():
        print_status("orchestrator.py not found", "error")
        return False
    
    print_status("orchestrator.py found", "success")
    
    # Test with a simple prompt
    test_command = [
        "python", "multi_claude.py",
        "-n", "2",
        "-p", "Test: Write a simple hello world function",
        "--provider", "qwen",
        "--debug"
    ]
    
    print_status("Testing multi-agent launch with Qwen provider...", "info")
    print_status(f"Command: {' '.join(test_command)}", "info")
    
    try:
        # Run for a short time to test
        process = subprocess.Popen(
            test_command,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
            text=True
        )
        
        # Wait for 10 seconds
        time.sleep(10)
        
        # Check if process is still running
        if process.poll() is None:
            print_status("Multi-agent orchestration started successfully", "success")
            
            # Terminate the test
            process.terminate()
            time.sleep(2)
            
            # Clean up tmux session
            subprocess.run(
                "tmux kill-session -t claude_agents",
                shell=True,
                capture_output=True
            )
            
            return True
        else:
            stdout, stderr = process.communicate()
            print_status("Multi-agent orchestration failed to start", "error")
            if stderr:
                print_status(f"Error: {stderr[:200]}", "error")
            return False
            
    except Exception as e:
        print_status(f"Failed to test multi-agent: {e}", "error")
        return False

def main():
    """Run all tests"""
    print(f"\n{BOLD}{'='*60}{RESET}")
    print(f"{BOLD}MaiFarm Qwen3-Coder 480B Integration Test Suite{RESET}")
    print(f"{BOLD}{'='*60}{RESET}")
    
    results = {}
    
    # Check environment
    results["environment"] = check_environment()
    
    # Test LLM proxy
    results["llm_proxy"] = test_llm_proxy()
    
    # Test Qwen API if proxy is available
    if results["llm_proxy"]:
        results["qwen_api"] = test_qwen_api()
    else:
        print_status("Skipping Qwen API test (proxy not available)", "warning")
        results["qwen_api"] = False
    
    # Test local Ollama
    results["ollama"] = test_ollama_local()
    
    # Test multi-agent if any provider is available
    if results["llm_proxy"] or results["ollama"]:
        results["multi_agent"] = test_multi_agent()
    else:
        print_status("Skipping multi-agent test (no providers available)", "warning")
        results["multi_agent"] = False
    
    # Summary
    print(f"\n{BOLD}=== Test Results Summary ==={RESET}")
    
    passed = sum(1 for v in results.values() if v)
    total = len(results)
    
    for test, result in results.items():
        status = "success" if result else "error"
        test_name = test.replace("_", " ").title()
        print_status(f"{test_name}: {'PASSED' if result else 'FAILED'}", status)
    
    print(f"\n{BOLD}Overall: {passed}/{total} tests passed{RESET}")
    
    if passed == total:
        print_status("All tests passed! Qwen3-Coder integration is ready.", "success")
        return 0
    elif passed > 0:
        print_status("Some tests passed. Partial functionality available.", "warning")
        return 1
    else:
        print_status("All tests failed. Please check configuration.", "error")
        return 2

if __name__ == "__main__":
    sys.exit(main())

File: tests/test-qwen-integration.js
#!/usr/bin/env node

/**
 * Comprehensive Qwen Integration Test Suite for MaiFarm
 * Tests farm creation, quick tasks, go wild mode, and WebSocket monitoring
 * 
 * Step 4: Test the integration for agent farms, go wild and quick tasks
 * Step 6: Handle gotchas and optimize
 */

const axios = require('axios');
const io = require('socket.io-client');
const { exec } = require('child_process');
const { promisify } = require('util');
const execAsync = promisify(exec);

// Configuration
const API_BASE_URL = process.env.API_URL || 'http://localhost:4567';
const WS_URL = process.env.WS_URL || 'http://localhost:4567';
const PROXY_URL = process.env.LLM_PROXY_URL || 'http://localhost:8001';

// Test results tracking
const testResults = {
  passed: [],
  failed: [],
  warnings: []
};

// ANSI color codes for output
const colors = {
  reset: '\x1b[0m',
  bright: '\x1b[1m',
  green: '\x1b[32m',
  red: '\x1b[31m',
  yellow: '\x1b[33m',
  blue: '\x1b[34m',
  cyan: '\x1b[36m'
};

// Logging helpers
function log(message, color = colors.reset) {
  console.log(`${color}${message}${colors.reset}`);
}

function logSection(title) {
  console.log('\n' + '='.repeat(60));
  log(title, colors.bright + colors.blue);
  console.log('='.repeat(60));
}

function logTest(name, status, details = '') {
  const statusSymbol = status === 'pass' ? '✓' : status === 'fail' ? '✗' : '⚠';
  const statusColor = status === 'pass' ? colors.green : status === 'fail' ? colors.red : colors.yellow;
  log(`${statusSymbol} ${name}`, statusColor);
  if (details) {
    console.log(`  ${details}`);
  }
}

// Test 1: Verify LLM Proxy is running and Qwen provider is available
async function testProxyAvailability() {
  logSection('Test 1: LLM Proxy and Qwen Provider Availability');
  
  try {
    // Check if proxy is running
    const proxyResponse = await axios.get(`${PROXY_URL}/providers`);
    
    if (proxyResponse.data.providers.includes('qwen')) {
      logTest('LLM Proxy running', 'pass', `Available providers: ${proxyResponse.data.providers.join(', ')}`);
      testResults.passed.push('LLM Proxy availability');
      
      // Check if Ollama is also available for local Qwen
      if (proxyResponse.data.providers.includes('ollama')) {
        logTest('Ollama provider available', 'pass', 'Local Qwen models can be used');
        testResults.passed.push('Ollama provider availability');
      } else {
        logTest('Ollama provider not available', 'warning', 'Only remote Qwen API will be tested');
        testResults.warnings.push('Ollama not available');
      }
      
      return true;
    } else {
      logTest('Qwen provider not available', 'fail', 'Qwen provider not found in proxy');
      testResults.failed.push('Qwen provider availability');
      return false;
    }
  } catch (error) {
    logTest('LLM Proxy connection failed', 'fail', error.message);
    log('  Please start the proxy: python llm_proxy.py --proxy', colors.yellow);
    testResults.failed.push('LLM Proxy connection');
    return false;
  }
}

// Test 2: Test Qwen API directly through proxy
async function testQwenAPI() {
  logSection('Test 2: Direct Qwen API Test via Proxy');
  
  try {
    const testMessage = {
      model: 'qwen/qwen-coder-480b',
      messages: [
        { role: 'system', content: 'You are a helpful coding assistant.' },
        { role: 'user', content: 'Write a simple Python function to calculate factorial.' }
      ],
      temperature: 0.7,
      max_tokens: 500
    };
    
    const response = await axios.post(`${PROXY_URL}/chat/completions`, testMessage);
    
    if (response.data.choices && response.data.choices[0].message.content) {
      const content = response.data.choices[0].message.content;
      logTest('Qwen API response received', 'pass', `Response length: ${content.length} chars`);
      
      // Check if response contains Python code
      if (content.includes('def') || content.includes('factorial')) {
        logTest('Response contains relevant code', 'pass');
        testResults.passed.push('Qwen API test');
      } else {
        logTest('Response may not contain expected code', 'warning');
        testResults.warnings.push('Qwen response quality');
      }
      
      // Log token usage if available
      if (response.data.usage) {
        log(`  Token usage: ${JSON.stringify(response.data.usage)}`, colors.cyan);
      }
      
      return true;
    } else {
      logTest('Invalid Qwen API response format', 'fail');
      testResults.failed.push('Qwen API response');
      return false;
    }
  } catch (error) {
    logTest('Qwen API call failed', 'fail', error.message);
    if (error.response?.data) {
      log(`  Error details: ${JSON.stringify(error.response.data)}`, colors.red);
    }
    testResults.failed.push('Qwen API call');
    return false;
  }
}

// Test 3: Test Farm Creation with Qwen
async function testFarmCreation() {
  logSection('Test 3: Farm Creation with Qwen Provider');
  
  try {
    const farmConfig = {
      name: 'test-qwen-farm-' + Date.now(),
      description: 'Test farm using Qwen3-Coder',
      agents: [
        {
          name: 'qwen-agent-1',
          role: 'developer',
          model: 'qwen/qwen-coder-480b',
          capabilities: ['coding', 'testing']
        },
        {
          name: 'qwen-agent-2',
          role: 'reviewer',
          model: 'qwen/qwen-coder-480b',
          capabilities: ['review', 'documentation']
        }
      ],
      task: 'Create a simple REST API with two endpoints',
      provider: 'qwen'
    };
    
    const response = await axios.post(`${API_BASE_URL}/api/farms/create`, farmConfig);
    
    if (response.data.success && response.data.farmId) {
      logTest('Farm created successfully', 'pass', `Farm ID: ${response.data.farmId}`);
      testResults.passed.push('Farm creation');
      
      // Monitor farm status
      const farmId = response.data.farmId;
      await monitorFarmStatus(farmId);
      
      return farmId;
    } else {
      logTest('Farm creation failed', 'fail', 'Invalid response');
      testResults.failed.push('Farm creation');
      return null;
    }
  } catch (error) {
    logTest('Farm creation error', 'fail', error.message);
    testResults.failed.push('Farm creation');
    return null;
  }
}

// Test 4: WebSocket Real-time Monitoring
async function monitorFarmStatus(farmId) {
  return new Promise((resolve) => {
    logSection('Test 4: WebSocket Real-time Monitoring');
    
    const socket = io(WS_URL, {
      transports: ['websocket'],
      reconnection: true
    });
    
    const timeout = setTimeout(() => {
      logTest('WebSocket monitoring timeout', 'warning', 'No updates received in 30s');
      testResults.warnings.push('WebSocket timeout');
      socket.disconnect();
      resolve();
    }, 30000);
    
    socket.on('connect', () => {
      logTest('WebSocket connected', 'pass');
      socket.emit('subscribe:farm', { farmId });
    });
    
    socket.on('farm:status', (data) => {
      log(`  Farm status: ${data.status}`, colors.cyan);
      if (data.agents) {
        data.agents.forEach(agent => {
          log(`    Agent ${agent.name}: ${agent.status}`, colors.cyan);
        });
      }
    });
    
    socket.on('task:progress', (data) => {
      log(`  Task progress: ${data.progress}%`, colors.cyan);
      if (data.progress === 100) {
        logTest('Task completed', 'pass');
        testResults.passed.push('WebSocket monitoring');
        clearTimeout(timeout);
        socket.disconnect();
        resolve();
      }
    });
    
    socket.on('error', (error) => {
      logTest('WebSocket error', 'fail', error.message);
      testResults.failed.push('WebSocket monitoring');
      clearTimeout(timeout);
      socket.disconnect();
      resolve();
    });
  });
}

// Test 5: Quick Task with Qwen
async function testQuickTask() {
  logSection('Test 5: Quick Task Execution with Qwen');
  
  try {
    const quickTask = {
      prompt: 'Write a TypeScript function to validate email addresses',
      provider: 'qwen',
      model: 'qwen/qwen-coder-480b',
      maxTokens: 1000,
      temperature: 0.5
    };
    
    const response = await axios.post(`${API_BASE_URL}/api/quicktask/execute`, quickTask);
    
    if (response.data.success && response.data.result) {
      logTest('Quick task executed', 'pass');
      
      // Validate the response contains TypeScript code
      const result = response.data.result;
      if (result.includes('function') || result.includes('const') || result.includes('email')) {
        logTest('Response contains relevant code', 'pass');
        testResults.passed.push('Quick task execution');
      } else {
        logTest('Response quality check', 'warning', 'May not contain expected code');
        testResults.warnings.push('Quick task quality');
      }
      
      // Check execution time
      if (response.data.executionTime) {
        log(`  Execution time: ${response.data.executionTime}ms`, colors.cyan);
        if (response.data.executionTime > 10000) {
          logTest('Performance warning', 'warning', 'Task took over 10 seconds');
          testResults.warnings.push('Quick task performance');
        }
      }
      
      return true;
    } else {
      logTest('Quick task failed', 'fail');
      testResults.failed.push('Quick task execution');
      return false;
    }
  } catch (error) {
    logTest('Quick task error', 'fail', error.message);
    testResults.failed.push('Quick task execution');
    return false;
  }
}

// Test 6: Go Wild Mode with Qwen
async function testGoWildMode() {
  logSection('Test 6: Go Wild Mode with Qwen');
  
  try {
    const goWildConfig = {
      initialPrompt: 'Explore and generate ideas for a modern web application',
      provider: 'qwen',
      model: 'qwen/qwen-coder-480b',
      maxIterations: 3,
      creativity: 0.8,
      contextWindow: 100000 // Leverage Qwen's large context window
    };
    
    const response = await axios.post(`${API_BASE_URL}/api/gowild/start`, goWildConfig);
    
    if (response.data.success && response.data.sessionId) {
      logTest('Go Wild session started', 'pass', `Session ID: ${response.data.sessionId}`);
      
      // Monitor Go Wild progress
      await monitorGoWildSession(response.data.sessionId);
      
      testResults.passed.push('Go Wild mode');
      return true;
    } else {
      logTest('Go Wild mode failed to start', 'fail');
      testResults.failed.push('Go Wild mode');
      return false;
    }
  } catch (error) {
    logTest('Go Wild mode error', 'fail', error.message);
    testResults.failed.push('Go Wild mode');
    return false;
  }
}

// Monitor Go Wild Session
async function monitorGoWildSession(sessionId) {
  return new Promise((resolve) => {
    const checkInterval = setInterval(async () => {
      try {
        const response = await axios.get(`${API_BASE_URL}/api/gowild/status/${sessionId}`);
        
        if (response.data.status === 'completed') {
          logTest('Go Wild session completed', 'pass');
          
          if (response.data.results) {
            log(`  Generated ${response.data.results.length} ideas`, colors.cyan);
          }
          
          clearInterval(checkInterval);
          resolve();
        } else if (response.data.status === 'failed') {
          logTest('Go Wild session failed', 'fail');
          clearInterval(checkInterval);
          resolve();
        } else {
          log(`  Go Wild status: ${response.data.status}`, colors.cyan);
        }
      } catch (error) {
        logTest('Go Wild monitoring error', 'fail', error.message);
        clearInterval(checkInterval);
        resolve();
      }
    }, 2000);
    
    // Timeout after 30 seconds
    setTimeout(() => {
      clearInterval(checkInterval);
      logTest('Go Wild timeout', 'warning', 'Session did not complete in 30s');
      resolve();
    }, 30000);
  });
}

// Test 7: Context Window Management
async function testContextWindowManagement() {
  logSection('Test 7: Large Context Window Management (256K tokens)');
  
  try {
    // Create a large context by concatenating multiple code files
    const largeContext = `
      // This is a test of Qwen's 256K context window capability
      ${Array(100).fill(null).map((_, i) => `
        function testFunction${i}() {
          // Function ${i} implementation
          const data = Array(1000).fill(${i});
          return data.reduce((a, b) => a + b, 0);
        }
      `).join('\n')}
    `;
    
    const request = {
      model: 'qwen/qwen-coder-480b',
      messages: [
        { role: 'system', content: 'You are analyzing a large codebase.' },
        { role: 'user', content: `Analyze this code and identify patterns:\n${largeContext}` }
      ],
      max_tokens: 2000
    };
    
    const startTime = Date.now();
    const response = await axios.post(`${PROXY_URL}/chat/completions`, request);
    const endTime = Date.now();
    
    if (response.data.choices && response.data.choices[0].message.content) {
      const executionTime = endTime - startTime;
      logTest('Large context processed', 'pass', `Time: ${executionTime}ms`);
      
      if (response.data.usage) {
        log(`  Tokens used: ${JSON.stringify(response.data.usage)}`, colors.cyan);
        
        if (response.data.usage.prompt_tokens > 50000) {
          logTest('Successfully handled large context', 'pass', 
                 `Processed ${response.data.usage.prompt_tokens} prompt tokens`);
          testResults.passed.push('Large context handling');
        }
      }
      
      return true;
    } else {
      logTest('Large context processing failed', 'fail');
      testResults.failed.push('Large context handling');
      return false;
    }
  } catch (error) {
    logTest('Context window test error', 'fail', error.message);
    testResults.failed.push('Large context handling');
    return false;
  }
}

// Test 8: Error Handling and Fallback
async function testErrorHandlingAndFallback() {
  logSection('Test 8: Error Handling and Fallback Mechanisms');
  
  try {
    // Test with invalid API key
    const invalidRequest = {
      model: 'qwen/invalid-model',
      messages: [{ role: 'user', content: 'test' }]
    };
    
    try {
      await axios.post(`${PROXY_URL}/chat/completions`, invalidRequest);
      logTest('Invalid model should have failed', 'fail');
      testResults.failed.push('Error handling');
    } catch (error) {
      if (error.response && error.response.status === 400) {
        logTest('Invalid model properly rejected', 'pass');
        testResults.passed.push('Error handling');
      } else {
        logTest('Unexpected error response', 'fail');
        testResults.failed.push('Error handling');
      }
    }
    
    // Test fallback to Ollama if Qwen API fails
    const fallbackRequest = {
      model: 'ollama/qwen2.5-coder:7b',
      messages: [{ role: 'user', content: 'Simple test' }],
      fallback: true
    };
    
    try {
      const response = await axios.post(`${PROXY_URL}/chat/completions`, fallbackRequest);
      if (response.data.choices) {
        logTest('Fallback to Ollama successful', 'pass');
        testResults.passed.push('Fallback mechanism');
      }
    } catch (error) {
      logTest('Fallback test skipped', 'warning', 'Ollama not available');
      testResults.warnings.push('Fallback test');
    }
    
    return true;
  } catch (error) {
    logTest('Error handling test failed', 'fail', error.message);
    testResults.failed.push('Error handling test');
    return false;
  }
}

// Test 9: Performance Optimization Tests
async function testPerformanceOptimizations() {
  logSection('Test 9: Performance Optimizations');
  
  const performanceTests = [];
  
  // Test response caching
  try {
    const testMessage = {
      model: 'qwen/qwen-coder-480b',
      messages: [{ role: 'user', content: 'What is 2+2?' }],
      temperature: 0 // Deterministic for caching
    };
    
    // First call
    const start1 = Date.now();
    await axios.post(`${PROXY_URL}/chat/completions`, testMessage);
    const time1 = Date.now() - start1;
    
    // Second call (should be cached)
    const start2 = Date.now();
    await axios.post(`${PROXY_URL}/chat/completions`, testMessage);
    const time2 = Date.now() - start2;
    
    if (time2 < time1 * 0.5) {
      logTest('Response caching working', 'pass', `Cached: ${time2}ms vs Original: ${time1}ms`);
      testResults.passed.push('Response caching');
    } else {
      logTest('Response caching', 'warning', 'Cache may not be working');
      testResults.warnings.push('Response caching');
    }
  } catch (error) {
    logTest('Caching test error', 'fail', error.message);
    testResults.failed.push('Response caching');
  }
  
  // Test batch processing
  try {
    const batchRequests = Array(5).fill(null).map((_, i) => ({
      model: 'qwen/qwen-coder-480b',
      messages: [{ role: 'user', content: `Test message ${i}` }],
      max_tokens: 100
    }));
    
    const startBatch = Date.now();
    const batchPromises = batchRequests.map(req => 
      axios.post(`${PROXY_URL}/chat/completions`, req)
    );
    
    const results = await Promise.all(batchPromises);
    const batchTime = Date.now() - startBatch;
    
    if (results.every(r => r.data.choices)) {
      logTest('Batch processing successful', 'pass', 
             `Processed ${results.length} requests in ${batchTime}ms`);
      testResults.passed.push('Batch processing');
      
      const avgTime = batchTime / results.length;
      log(`  Average time per request: ${avgTime.toFixed(0)}ms`, colors.cyan);
    } else {
      logTest('Batch processing incomplete', 'fail');
      testResults.failed.push('Batch processing');
    }
  } catch (error) {
    logTest('Batch processing error', 'fail', error.message);
    testResults.failed.push('Batch processing');
  }
  
  return true;
}

// Main test runner
async function runAllTests() {
  console.log(colors.bright + colors.cyan);
  console.log('╔════════════════════════════════════════════════════════════╗');
  console.log('║     Qwen3-Coder Integration Test Suite for MaiFarm        ║');
  console.log('║                    Step 4 & Step 6                        ║');
  console.log('╚════════════════════════════════════════════════════════════╝');
  console.log(colors.reset);
  
  // Run tests in sequence
  const proxyAvailable = await testProxyAvailability();
  
  if (!proxyAvailable) {
    log('\nCannot proceed without LLM Proxy. Please start it first:', colors.red);
    log('python llm_proxy.py --proxy', colors.yellow);
    process.exit(1);
  }
  
  await testQwenAPI();
  
  const farmId = await testFarmCreation();
  
  await testQuickTask();
  
  await testGoWildMode();
  
  await testContextWindowManagement();
  
  await testErrorHandlingAndFallback();
  
  await testPerformanceOptimizations();
  
  // Print summary
  logSection('Test Summary');
  
  const totalTests = testResults.passed.length + testResults.failed.length;
  const passRate = totalTests > 0 ? 
    ((testResults.passed.length / totalTests) * 100).toFixed(1) : 0;
  
  log(`Total Tests: ${totalTests}`, colors.bright);
  log(`Passed: ${testResults.passed.length}`, colors.green);
  log(`Failed: ${testResults.failed.length}`, colors.red);
  log(`Warnings: ${testResults.warnings.length}`, colors.yellow);
  log(`Pass Rate: ${passRate}%`, passRate >= 80 ? colors.green : colors.yellow);
  
  if (testResults.failed.length > 0) {
    log('\nFailed Tests:', colors.red);
    testResults.failed.forEach(test => log(`  - ${test}`, colors.red));
  }
  
  if (testResults.warnings.length > 0) {
    log('\nWarnings:', colors.yellow);
    testResults.warnings.forEach(warning => log(`  - ${warning}`, colors.yellow));
  }
  
  // Exit with appropriate code
  process.exit(testResults.failed.length > 0 ? 1 : 0);
}

// Handle errors gracefully
process.on('unhandledRejection', (error) => {
  console.error('Unhandled rejection:', error);
  process.exit(1);
});

// Run the tests
runAllTests().catch(error => {
  console.error('Test suite error:', error);
  process.exit(1);
});